### ISSUE TITLE
ALX: a function that prints anything (variadic function)

### ISSUE THREAD

**Issue by ahmed149** [ahmed149/learning-programing-in-C#75]
#include "variadic_functions.h"
#include <stdarg.h>
#include <stdio.h>

/**
 * _printchar - function to print character
 * @list: list being passed
 *
 * Return: void
 */
void _printchar(va_list list)
{
	printf("%c", va_arg(list, int));
}

/**
 * _printstr - function to print string
 * @list: list being passed
 * Return: void
 *
 */
void _printstr(va_list list)
{
	char *s;

	s = va_arg(list, char *);
	if (s == 0)
		s = "(nil)";
	printf("%s", s);
}

/**
 * _printfloat - function to print float
 * @list: list being passed
 * Return: void
 *
 */
void _printfloat(va_list list)
{
	printf("%f", va_arg(list, double));
}

/**
 * _printint - function to print int
 * @list: list being passed
 * Return: void
 *
 */
void _printint(va_list list)
{
	printf("%d", va_arg(list, int));
}

/**
 * print_all - function to print various arguments
 * @format: constant char argument
 * Return: Void
 */
void print_all(const char * const format, ...)
{
	unsigned int a = 0, b = 0;
	va_list list;
	char *str = "";

	what_format frmt[] = {
		{"c", _printchar},
		{"f", _printfloat},
		{"s", _printstr},
		{"i", _printint},
	};

	va_start(list, format);
	while (format != 0 && format[a / 4] != 0)
	{
		b = a % 4;
		if (frmt[b].type[0] == format[a / 4])
		{
			printf("%s", str);
			frmt[b].f(list);
			str = ", ";
		}
		a++;
	}
	printf("\n");
	va_end(list);
}

**Comment by ahmed149** [comment]
This C code defines a variadic function called `print_all`, which is designed to print a list of various types of arguments based on a format string. It uses the `<stdarg.h>` library to work with variable-length argument lists. Additionally, it defines several helper functions (`_printchar`, `_printfloat`, `_printstr`, `_printint`) to handle specific data types. Let's break down the code:

1. `#include "variadic_functions.h"` and `#include <stdarg.h>`:
   - These lines include a user-defined header file, "variadic_functions.h," and the standard C library `<stdarg.h>`. The latter is necessary for working with variable-length argument lists.

2. `#include <stdio.h>`:
   - This line includes the standard C library `<stdio.h>` for input and output operations since the functions print data to the console.

3. Helper Functions:
   - The code defines four helper functions, `_printchar`, `_printfloat`, `_printstr`, and `_printint`. Each function corresponds to a specific data type and is responsible for printing that type of data. They all take a `va_list` as an argument, which allows them to retrieve and print the next argument of the respective data type.

4. `void print_all(const char * const format, ...)`:
   - This is the main function, `print_all`, that accepts a format string and a variable number of arguments. It will print the provided arguments according to the format string.

5. `unsigned int a = 0, b = 0;`, `va_list list;`, `char *str = "";`:
   - These lines declare variables for use within the `print_all` function. `a` and `b` are loop counters, `list` is a `va_list` used to manage the variable-length argument list, and `str` is used to handle the string separator between printed values.

6. `what_format frmt[] = { ... }`:
   - This is an array of structures that associates format specifiers with their respective printing functions. For example, "c" corresponds to `_printchar`.

7. `va_start(list, format);`:
   - This initializes the `list` variable to point to the first unnamed argument after `format`, preparing to traverse the variable-length argument list.

8. `while (format != 0 && format[a / 4] != 0)`:
   - This loop iterates through the `format` string and processes each character. It stops when it reaches the end of the string (`format[a / 4] != 0`).

9. `b = a % 4;`:
   - This calculates the index in the `frmt` array based on the current position in the `format` string (0 to 3 since there are four format specifiers).

10. `if (frmt[b].type[0] == format[a / 4])`:
    - This condition checks if the current format character matches one of the format specifiers in the `frmt` array.

11. `printf("%s", str);`:
    - This prints the separator `str` between values, ensuring proper formatting between multiple printed values.

12. `frmt[b].f(list);`:
    - This calls the appropriate printing function from the `frmt` array based on the format specifier encountered.

13. `str = ", ";`:
    - This sets `str` to the separator string for subsequent iterations.

14. `a++;`:
    - Increment the loop counter to move to the next character in the `format` string.

15. `printf("\n");`:
    - Finally, after processing all the arguments, the code prints a newline character to end the line of output.

16. `va_end(list);`:
    - This cleans up the `list` variable, signaling that you've finished working with the variable-length argument list.

In summary, the `print_all` function allows you to pass a format string along with a variable number of arguments of different types. It prints these arguments according to the format specified in the string, using the helper functions for each data type. The format string is processed character by character, and the corresponding printing function is called for each argument. It also handles separators between printed values and NULL strings appropriately.
